### DISCUSSION TITLE
Is there any way to subscribe to the event of bucket refill?

### DISCUSSION THREAD

**Discussion by DAv10195** [vladimir-bukhtoyarov/bucket4j#224] [Q&A] [answered]
Hi,

My name is David and I'm using this great library to implement a token based throttling mechanism.
I've initialized my bucket in the following way to allow 3 tokens per 10 seconds, which are getting refreshed intervally:

`Bucket bucket = Bucket.builder().addLimit(Bandwidth.classic(3, Refill.intervally(3, Duration.ofSeconds(10)))).build();`

The general scenario I want to use this bucket in consists of a queue getting filled by some producer thread(s), which simultaneously getting consumed by some consumer thread(s).
Each time a consumer thread reads an element from the queue, it has to acquire a token from the above bucket to perform the required operation on that element (by invoking `bucket.tryConsume(1)`) and if there's no token currently available, then my current implementation sends the thread to sleep for some time.
When the thread wakes up, it runs again the flow above (if there's an element in the queue, try acquiring a token, if no token go to sleep, etc.).

As the scenario I described above involves some busy waiting (sleeping for some time), I wonder if there's some way to register/listen to the event of the bucket refill.
I did have a look at the BlockingBucket and BucketListener interfaces, but I couldn't figure out a way to use them to catch the moment of bucket refill.

Any ideas will be highly appreciated :)

Thanks in advance,
David

**Comment by DAv10195** [comment, marked as answer]
Looks like I found a way to avoid the busy waiting with the sleep time. When `bucket.tryConsume(1)` returns false, I run the following code:

`bucket.asBlocking().consume(1);
bucket.addTokens(1);`

Which leads to the consumer thread being blocked until the bucket get refilled. When it wakes up, it immediately returns the consumed token to the bucket.

What do you think about this solution? is it a recommended way to achieve this behavior?

Thanks again,
David

**Reply by vladimir-bukhtoyarov** [reply]
Hello @DAv10195

I do not understand one point: why are you going to use addTokens? From my opinion bucket.asBlocking().consume(1) is fully enough. What is the motivation to return tokens back?

**Reply by DAv10195** [reply]
yes it seems I don't really need to put it back... Thanks a lot for the quick response, highly appreciated :)
